concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -1226,24 +1250,73 @@ instance HashableTo BlockQuasiHash (BakedBlock pv) where
               where
                 transactionsAndOutcomesHash = Hash.hashOfHashes transactionsHash outcomesHash
                   where
-                    transactionsHash = computeTransactionsHash bbTransactions
+                    transactionsHash = computeTransactionsHash (sing @bhv) bbTransactions
                     outcomesHash = tohGet dbhv0TransactionOutcomesHash
                 stateHash = v0StateHash dbhv0BlockStateHash
             DerivableBlockHashesV1{..} -> Hash.hashOfHashes transactionsHash blockResultHash
               where
-                transactionsHash = computeTransactionsHash bbTransactions
+                transactionsHash = computeTransactionsHash (sing @bhv) bbTransactions
                 blockResultHash = theBlockResultHash dbhv1BlockResultHash
 
+instance (IsProtocolVersion pv) => HashableTo SuccessorProof (BakedBlock pv) where
+    getHash = makeSuccessorProof @(BlockHashVersionFor pv) . getHash
+
 -- | Compute the block hash from the header hash and quasi-hash.
-computeBlockHash :: BlockHeaderHash -> BlockQuasiHash -> BlockHash
-computeBlockHash bhh bqh =
+computeBlockHash' :: BlockHeaderHash -> Hash.Hash -> BlockHash
+{-# INLINE computeBlockHash' #-}
+computeBlockHash' bhh bqh =
     BlockHash $
         Hash.hashOfHashes
             (theBlockHeaderHash bhh)
-            (theBlockQuasiHash bqh)
+            bqh
 
-instance HashableTo BlockHash (BakedBlock pv) where
-    getHash bb = computeBlockHash (getHash bb) (getHash bb)
+-- | Compute the block hash from the header hash and quasi-hash.
+computeBlockHash :: BlockHeaderHash -> BlockQuasiHash' bhv -> BlockHash
+computeBlockHash bhh bqh = computeBlockHash' bhh (theBlockQuasiHash bqh)
+
+instance (IsProtocolVersion pv) => HashableTo BlockHash (BakedBlock pv) where
+    getHash bb = computeBlockHash @(BlockHashVersionFor pv) (getHash bb) (getHash bb)
+
+instance (Monad m, BlockHashVersionFor pv ~ 'BlockHashVersion1) => Merkle.MerkleProvable m (BakedBlock pv) where
+    buildMerkleProof open BakedBlock{..} = do
+        let blockHeader = optProof ["header"] . rawMerkle . runPut $ do
+                put bbRound
+                put bbEpoch
+                put (qcBlock bbQuorumCertificate)
+        let biPath = ["quasi", "meta", "bakerInfo"]
+        let timestampBaker = optProof (biPath ++ ["timestampBaker"]) . rawMerkle . runPut $ do
+                put bbTimestamp
+                put bbBaker
+        let nonce = optProof (biPath ++ ["nonce"]) . rawMerkle . encode $ bbNonce
+        let bakerInfo = optProof biPath [timestampBaker, nonce]
+        let chPath = ["quasi", "meta", "certificatesHash"]
+        let qcPath = chPath ++ ["quorumCertificate"]
+        qcMerkleProof <- Merkle.buildMerkleProof (open . (qcPath ++)) bbQuorumCertificate
+        let qcHash = optProof qcPath qcMerkleProof
+        let tfPath = chPath ++ ["timeoutFinalization"]
+        let tcPath = tfPath ++ ["timeoutCertificate"]
```

Comment:
Is it intended that the timeout certificate is under timeoutFinalization?

Reply:
Yes. timeoutFinalization combines the timeout certificate and epoch finalization entry.